substrate/frame/delegated-staking/src/tests.rs

```diff
@@ -0,0 +1,647 @@
+			// is 3 eras.
+			assert_eq!(get_agent(&agent).ledger.unclaimed_withdrawals, 8 * 10);
+
+			// release some delegation now.
+			assert_ok!(DelegatedStaking::release_delegation(
+				RawOrigin::Signed(agent).into(),
+				300,
+				40,
+				0
+			));
+			assert_eq!(get_agent(&agent).ledger.unclaimed_withdrawals, 80 - 40);
+
+			// cannot release more than available
+			assert_noop!(
+				DelegatedStaking::release_delegation(RawOrigin::Signed(agent).into(), 300, 50, 0),
+				Error::<T>::NotEnoughFunds
+			);
+			assert_ok!(DelegatedStaking::release_delegation(
+				RawOrigin::Signed(agent).into(),
+				300,
+				40,
+				0
+			));
+
+			assert_eq!(DelegatedStaking::held_balance_of(&300), 100 - 80);
+		});
+	}
+
+	#[test]
+	fn reward_destination_restrictions() {
+		ExtBuilder::default().build_and_execute(|| {
+			// give some funds to 200
+			fund(&200, 1000);
+			let balance_200 = Balances::free_balance(200);
+
+			// `Agent` account cannot be reward destination
+			assert_noop!(
+				DelegatedStaking::register_agent(RawOrigin::Signed(200).into(), 200),
+				Error::<T>::InvalidRewardDestination
+			);
+
+			// different reward account works
+			assert_ok!(DelegatedStaking::register_agent(RawOrigin::Signed(200).into(), 201));
+			// add some delegations to it
+			fund(&300, 1000);
+			assert_ok!(DelegatedStaking::delegate_to_agent(
+				RawOrigin::Signed(300).into(),
+				200,
+				100
+			));
+
+			// update_payee to self fails.
+			assert_noop!(
+				<Staking as StakingInterface>::update_payee(&200, &200),
+				StakingError::<T>::RewardDestinationRestricted
+			);
+
+			// passing correct reward destination works
+			assert_ok!(<Staking as StakingInterface>::update_payee(&200, &201));
+
+			// amount is staked correctly
+			assert!(eq_stake(200, 100, 100));
+			assert_eq!(get_agent(&200).available_to_bond(), 0);
+			assert_eq!(get_agent(&200).ledger.effective_balance(), 100);
+
+			// free balance of delegate is untouched
+			assert_eq!(Balances::free_balance(200), balance_200);
+		});
+	}
+
+	#[test]
+	fn agent_restrictions() {
+		ExtBuilder::default().build_and_execute(|| {
+			setup_delegation_stake(200, 201, (202..203).collect(), 100, 0);
+
+			// Registering again is noop
+			assert_noop!(
+				DelegatedStaking::register_agent(RawOrigin::Signed(200).into(), 201),
+				Error::<T>::NotAllowed
+			);
+			// a delegator cannot become delegate
+			assert_noop!(
+				DelegatedStaking::register_agent(RawOrigin::Signed(202).into(), 203),
+				Error::<T>::NotAllowed
+			);
+			// existing staker cannot become a delegate
+			assert_noop!(
+				DelegatedStaking::register_agent(
+					RawOrigin::Signed(GENESIS_NOMINATOR_ONE).into(),
+					201
+				),
+				Error::<T>::AlreadyStaking
+			);
+			assert_noop!(
+				DelegatedStaking::register_agent(RawOrigin::Signed(GENESIS_VALIDATOR).into(), 201),
+				Error::<T>::AlreadyStaking
+			);
+		});
+	}
+
+	#[test]
+	fn migration_works() {
```

Comment:
How about a set of live tests using `remote_externalities`? that we can then easily add to the fellowship runtimes and test things with? bags-list pallet has similar ones.

Reply:
Yeah I have been thinking about having a good set of remote externalities test for the whole integration. Probably will be part of next PR.